mooncake-store/include/allocation_strategy.h

```diff
@@ -1,138 +1,378 @@
- * specified, then falls back to random allocation among all available
- * allocators.
+ * This strategy ensures that for each slice, its replicas are placed in
+ * different segments. Different slices may use the same segments.
  */
 class RandomAllocationStrategy : public AllocationStrategy {
    public:
     RandomAllocationStrategy() : rng_(std::random_device{}()) {}
 
-    std::unique_ptr<AllocatedBuffer> Allocate(
+    AllocationResult Allocate(
         const std::vector<std::shared_ptr<BufferAllocatorBase>>& allocators,
         const std::unordered_map<
             std::string, std::vector<std::shared_ptr<BufferAllocatorBase>>>&
             allocators_by_name,
-        size_t objectSize, const ReplicateConfig& config) override {
-        // Fast path: single allocator case
-        if (allocators.size() == 1) {
-            return allocators[0]->allocate(objectSize);
+        const std::vector<size_t>& slice_sizes,
+        const ReplicateConfig& config) override {
+        AllocationResult result;
+
+        if (!validateInput(slice_sizes, config.replica_num, result)) {
+            return result;
+        }
+
+        std::vector<std::vector<std::unique_ptr<AllocatedBuffer>>>
+            replica_buffers(config.replica_num);
+        for (auto& replica_buffer : replica_buffers) {
+            replica_buffer.reserve(slice_sizes.size());
+        }
+
+        // Allocate each slice across replicas
+        for (size_t slice_idx = 0; slice_idx < slice_sizes.size();
+             ++slice_idx) {
+            auto slice_replicas = allocateSlice(allocators, allocators_by_name,
+                                                slice_sizes[slice_idx],
+                                                config.replica_num, config);
+
+            if (slice_replicas.size() != config.replica_num) {
+                result.status = AllocationResult::Status::FAILURE;
+                result.error_message = fmt::format(
+                    "Failed to allocate {} replicas for slice {} (size: {}). "
+                    "Only {} replicas can be allocated.",
+                    config.replica_num, slice_idx, slice_sizes[slice_idx],
+                    slice_replicas.size());
+                return result;
+            }
+
+            for (size_t replica_idx = 0; replica_idx < config.replica_num;
+                 ++replica_idx) {
+                replica_buffers[replica_idx].push_back(
+                    std::move(slice_replicas[replica_idx]));
+            }
+        }
+
+        result.replicas.reserve(config.replica_num);
+        for (size_t replica_idx = 0; replica_idx < config.replica_num;
+             ++replica_idx) {
+            result.replicas.emplace_back(
+                std::move(replica_buffers[replica_idx]),
+                ReplicaStatus::PROCESSING);
+        }
+        result.status = AllocationResult::Status::SUCCESS;
+
+        return result;
+    }
+
+    AllocationResult AllocateAdditionalReplicas(
```

Comment:
It appears we haven't use this func?  We could remove it to streamline the codebase and reduce complexity.

Reply:
removed